module/os/linux/zfs/zpl_file.c

```diff
@@ -743,21 +746,46 @@ zpl_fallocate_common(struct inode *ip, int mode, loff_t offset, loff_t len)
 	spl_inode_lock(ip);
 	olen = i_size_read(ip);
 
-	if (offset > olen) {
-		spl_inode_unlock(ip);
-		return (0);
-	}
-	if (offset + len > olen)
-		len = olen - offset;
-	bf.l_type = F_WRLCK;
-	bf.l_whence = SEEK_SET;
-	bf.l_start = offset;
-	bf.l_len = len;
-	bf.l_pid = 0;
-
 	crhold(cr);
 	cookie = spl_fstrans_mark();
-	error = -zfs_space(ITOZ(ip), F_FREESP, &bf, O_RDWR, offset, cr);
+	if (mode & FALLOC_FL_PUNCH_HOLE) {
+		flock64_t bf;
+
+		if (offset > olen)
+			goto out_unmark;
+
+		if (offset + len > olen)
+			len = olen - offset;
+		bf.l_type = F_WRLCK;
+		bf.l_whence = SEEK_SET;
+		bf.l_start = offset;
+		bf.l_len = len;
+		bf.l_pid = 0;
+
+		error = -zfs_space(ITOZ(ip), F_FREESP, &bf, O_RDWR, offset, cr);
+	} else if ((mode & ~FALLOC_FL_KEEP_SIZE) == 0) {
+		struct kstatfs statfs;
+
+		/*
+		 * Use zfs_statvfs() instead of dmu_objset_space() since it
+		 * also checks project quota limits, which are relevant here.
+		 */
+		error = zfs_statvfs(ip, &statfs);
+		if (error)
+			goto out_unmark;
+		/*
+		 * Allow for some extra space for overhead/races. Divide
+		 * first to avoid potential overflow. We know the product
+		 * previously fit into availbytes from dmu_objset_space().
+		 */
+		if (len > statfs.f_bsize / 8 * 7 * statfs.f_bavail) {
```

Comment:
No, the code is correct as-is I think?  This is *decreasing* the calculated free space by 1/8, rather than *increasing* the write size by 1/8 (i.e. not `len * 9 / 8`), to avoid overflowing if a very large `len` is passed.  While `f_bsize` is always going to be at least 512, there is no chance of the /8 resulting in a zero size, but if `f_bavail` = 0 then clearly there isn't enough space in the filesystem for the `fallocate()` to return success.

Reply:
My mistake, you're right.

Reply:
In closer review of the code, I realize that this could also have been written in the more normal "* 7 / 8" format since this would be applied to `f_bsize` first and could not overflow the intermediate result due to the maximum `f_bsize` being in the 1MiB range.  I'll fix that in the updated version incorporating your parameter.